karapace/protobuf/syntax_reader.py

```diff
@@ -0,0 +1,367 @@
+            while True:
+                self.expect(self.pos < len(self.data), "trailing comment must be closed")
+                if self.data[self.pos] == '*' and self.pos + 1 < len(self.data) and self.data[self.pos + 1] == '/':
+                    end = self.pos - 1  # The character before '*'.
+                    self.pos += 2  # Skip to the character after '/'.
+                    break
+                self.pos += 1
+            # Ensure nothing follows a trailing star comment.
+            while self.pos < len(self.data):
+                c = self.data[self.pos]
+                self.pos += 1
+                if c == "\n":
+                    self.newline()
+                    break
+
+                self.expect(c in [" ", "\t"], "no syntax may follow trailing comment")
+
+        else:
+            # Consume comment until newline.
+            while True:
+                if self.pos == len(self.data):
+                    end = self.pos - 1
+                    break
+                c = self.data[self.pos]
+                self.pos += 1
+                if c == "\n":
+                    self.newline()
+                    end = self.pos - 2  # Account for stepping past the newline.
+                    break
+
+        # Remove trailing whitespace.
+        while end > start and (self.data[end] == " " or self.data[end] == "\t"):
+            end -= 1
+
+        if end == start:
+            return documentation
+
+        trailing_documentation = self.data[start:end + 1]
+        if not documentation.strip():
+            return trailing_documentation
+        return f"{documentation}\n{trailing_documentation}"
+
+    def skip_whitespace(self, skip_comments: bool) -> None:
+        """ Skips whitespace characters and optionally comments. When this returns, either
+        self.pos == self.data.length or a non-whitespace character.
+        """
+        while self.pos < len(self.data):
+            c = self.data[self.pos]
+            if c in [" ", "\t", "\r", "\n"]:
+                self.pos += 1
+                if c == "\n":
+                    self.newline()
+            elif skip_comments and c == "/":
+                self.read_comment()
+            else:
+                return None
+
+    def newline(self) -> None:
+        """ Call this every time a '\n' is encountered. """
+        self.line += 1
+        self.line_start = self.pos
+
+    def location(self) -> Location:
+        return self._location.at(self.line + 1, self.pos - self.line_start + 1)
+
+    def expect(self, condition: bool, message: str) -> None:
+        location = self.location()
+        if not condition:
+            self.unexpected(message, location)
+
+    def expect_with_location(self, condition: bool, location: Location, message: str) -> None:
+        if not condition:
+            self.unexpected(message, location)
+
+    def unexpected(self, message: str, location: Location = None) -> None:
```

Comment:
https://mypy.readthedocs.io/en/stable/more_types.html#the-noreturn-type

```suggestion
    def unexpected(self, message: str, location: Location = None) -> NoReturn:
```